File: lib/microsandbox/image.rb
# frozen_string_literal: true

module Microsandbox
  # Metadata for a cached OCI image, from {Image.get} / {Image.list}.
  class ImageInfo
    attr_reader :reference, :size_bytes, :manifest_digest, :architecture, :os, :layer_count

    def initialize(data)
      @reference = data["reference"]
      @size_bytes = data["size_bytes"]
      @manifest_digest = data["manifest_digest"]
      @architecture = data["architecture"]
      @os = data["os"]
      @layer_count = data["layer_count"]
      @created_at_ms = data["created_at_ms"]
      @last_used_at_ms = data["last_used_at_ms"]
    end

    # @return [Time, nil]
    def created_at
      @created_at_ms && Time.at(@created_at_ms / 1000.0)
    end

    # @return [Time, nil]
    def last_used_at
      @last_used_at_ms && Time.at(@last_used_at_ms / 1000.0)
    end

    def inspect
      "#<Microsandbox::ImageInfo reference=#{@reference.inspect} layers=#{@layer_count}>"
    end
  end

  # Full inspection detail for a cached image, from {Image.inspect}.
  class ImageDetail
    # @return [ImageInfo]
    attr_reader :handle
    # @return [Hash, nil] OCI config (digest, env, cmd, entrypoint, working_dir,
    #   user, labels, stop_signal). `labels` is a Hash (or nil) of OCI config labels.
    attr_reader :config
    # @return [Array<Hash>] layer descriptors
    attr_reader :layers

    def initialize(data)
      @handle = ImageInfo.new(data["handle"])
      @config = data["config"]
      @layers = data["layers"] || []
    end

    def reference = @handle.reference

    def inspect
      "#<Microsandbox::ImageDetail reference=#{@handle.reference.inspect} layers=#{@layers.size}>"
    end
  end

  # The result of {Image.prune}.
  class ImagePruneReport
    attr_reader :image_refs_removed, :manifests_removed, :layers_removed,
      :fsmeta_removed, :vmdk_removed, :bytes_reclaimed

    def initialize(data)
      @image_refs_removed = data["image_refs_removed"]
      @manifests_removed = data["manifests_removed"]
      @layers_removed = data["layers_removed"]
      @fsmeta_removed = data["fsmeta_removed"]
      @vmdk_removed = data["vmdk_removed"]
      @bytes_reclaimed = data["bytes_reclaimed"]
    end
  end

  # Management of the local OCI image cache. Images are pulled automatically by
  # {Sandbox.create}; this namespace lets you inspect and prune the cache.
  class Image
    class << self
      # All cached images.
      # @return [Array<ImageInfo>]
      def list
        Native::Image.list.map { |info| ImageInfo.new(info) }
      end

      # Metadata for one cached image.
      # @return [ImageInfo]
      def get(reference)
        ImageInfo.new(Native::Image.get(reference.to_s))
      end

      # Full inspection detail for a cached image. With no argument this is the
      # normal class `#inspect` (so object display still works).
      # @return [ImageDetail]
      def inspect(reference = nil)
        return super() if reference.nil?

        ImageDetail.new(Native::Image.inspect(reference.to_s))
      end

      # Remove a cached image.
      # @param force [Boolean] remove even if referenced
      # @return [nil]
      def remove(reference, force: false)
        Native::Image.remove(reference.to_s, force)
        nil
      end

      # Garbage-collect unreferenced images, manifests, and layers.
      # @return [ImagePruneReport]
      def prune
        ImagePruneReport.new(Native::Image.prune)
      end

      # Load images into the cache from a local `docker save` tarball or an
      # OCI Image Layout archive. Pass `"-"` to read the archive from `$stdin`
      # (spooled to a temp file first, like the Python SDK — the core reads
      # seekable files only). Mirrors the Python `Image.load` / Node
      # `imageLoad` (runtime v0.6.7).
      # @param input_path [String] archive path (or "-")
      # @param tag [String, Array<String>, nil] retag the loaded image(s)
      # @return [Array<ImageInfo>] one entry per loaded image
      def load(input_path, tag: nil)
        tags = Array(tag).map(&:to_s)
        path = input_path.to_s
        if path == "-"
          require "tempfile"
          Tempfile.create(["msb-image-load", ".tar"]) do |tmp|
            tmp.binmode
            IO.copy_stream($stdin, tmp)
            tmp.flush
            return Native::Image.load(tmp.path, tags).map { |info| ImageInfo.new(info) }
          end
        end
        Native::Image.load(path, tags).map { |info| ImageInfo.new(info) }
      end

      # Save cached image(s) into a local archive.
      # Mirrors the Python `Image.save` / Node `imageSave` (runtime v0.6.7).
      # @param reference [String, Array<String>] image reference(s) to save
      # @param output_path [String] destination archive path
      # @param format [:docker, :oci] archive layout (default :docker, the
      #   shape `docker load` expects)
      # @return [nil]
      def save(reference, output_path:, format: :docker)
        refs = Array(reference).map(&:to_s)
        raise ArgumentError, "at least one image reference is required" if refs.empty?

        Native::Image.save(refs, output_path.to_s, format.to_s)
        nil
      end
    end
  end
end

File: ext/microsandbox/src/image.rs
//! OCI image-cache management: `Microsandbox::Native::Image`.
//!
//! Mirrors `sdk/python/src/image.rs`. Static async operations over the cached
//! image store; results are returned as plain Ruby Hashes/Arrays. Note the core
//! `ImageHandle` exposes accessor *methods* while `ImageDetail`/`ImageConfigDetail`/
//! `ImageLayerDetail`/`ImagePruneReport` expose public *fields*.

use magnus::{function, prelude::*, Error, RArray, RHash, RModule, Ruby};
use microsandbox::image::{Image, ImageDetail, ImageHandle, ImagePruneReport};

use crate::backend::with_local_backend;
use crate::conv;
use crate::runtime::ruby;

fn handle_to_hash(h: &ImageHandle) -> RHash {
    let hash = ruby().hash_new();
    let _ = hash.aset("reference", h.reference().to_string());
    let _ = hash.aset("size_bytes", h.size_bytes());
    let _ = hash.aset("manifest_digest", h.manifest_digest().map(str::to_string));
    let _ = hash.aset("architecture", h.architecture().map(str::to_string));
    let _ = hash.aset("os", h.os().map(str::to_string));
    let _ = hash.aset("layer_count", h.layer_count());
    let _ = hash.aset(
        "created_at_ms",
        h.created_at().map(|dt| dt.timestamp_millis()),
    );
    let _ = hash.aset(
        "last_used_at_ms",
        h.last_used_at().map(|dt| dt.timestamp_millis()),
    );
    hash
}

fn detail_to_hash(detail: ImageDetail) -> RHash {
    let r = ruby();
    let hash = r.hash_new();
    let _ = hash.aset("handle", handle_to_hash(&detail.handle));

    if let Some(config) = detail.config {
        let c = r.hash_new();
        let _ = c.aset("digest", config.digest);
        let _ = c.aset("env", config.env);
        let _ = c.aset("cmd", config.cmd);
        let _ = c.aset("entrypoint", config.entrypoint);
        let _ = c.aset("working_dir", config.working_dir);
        let _ = c.aset("user", config.user);
        // OCI config labels: a free-form JSON object (or nil). Converted to a
        // Ruby Hash so `ImageDetail#config["labels"]` matches the Python/Node
        // `config.labels` dict. (The pinned v0.5.8 runtime persists this as
        // null today; the key exists for forward-compatibility/parity.)
        let _ = c.aset("labels", config.labels.as_ref().map(conv::json_to_ruby));
        let _ = c.aset("stop_signal", config.stop_signal);
        let _ = hash.aset("config", c);
    } else {
        let _ = hash.aset("config", r.qnil());
    }

    let layers = r.ary_new();
    for layer in detail.layers {
        let l = r.hash_new();
        let _ = l.aset("diff_id", layer.diff_id);
        let _ = l.aset("blob_digest", layer.blob_digest);
        let _ = l.aset("media_type", layer.media_type);
        let _ = l.aset("compressed_size_bytes", layer.compressed_size_bytes);
        let _ = l.aset("erofs_size_bytes", layer.erofs_size_bytes);
        let _ = l.aset("position", layer.position);
        let _ = layers.push(l);
    }
    let _ = hash.aset("layers", layers);
    hash
}

fn report_to_hash(report: ImagePruneReport) -> RHash {
    let hash = ruby().hash_new();
    let _ = hash.aset("image_refs_removed", report.image_refs_removed);
    let _ = hash.aset("manifests_removed", report.manifests_removed);
    let _ = hash.aset("layers_removed", report.layers_removed);
    let _ = hash.aset("fsmeta_removed", report.fsmeta_removed);
    let _ = hash.aset("vmdk_removed", report.vmdk_removed);
    let _ = hash.aset("bytes_reclaimed", report.bytes_reclaimed);
    hash
}

fn get(reference: String) -> Result<RHash, Error> {
    let handle = with_local_backend(async |local| Image::get_local(local, &reference).await)?;
    Ok(handle_to_hash(&handle))
}

fn list() -> Result<RArray, Error> {
    let handles = with_local_backend(async |local| Image::list_local(local).await)?;
    let arr = ruby().ary_new();
    for h in handles.iter() {
        arr.push(handle_to_hash(h))?;
    }
    Ok(arr)
}

fn inspect(reference: String) -> Result<RHash, Error> {
    let detail = with_local_backend(async |local| Image::inspect_local(local, &reference).await)?;
    Ok(detail_to_hash(detail))
}

fn remove(reference: String, force: bool) -> Result<(), Error> {
    with_local_backend(async |local| Image::remove_local(local, &reference, force).await)
}

fn prune() -> Result<RHash, Error> {
    let report = with_local_backend(async |local| Image::prune_local(local).await)?;
    Ok(report_to_hash(report))
}

pub fn define(ruby: &Ruby, native: &RModule) -> Result<(), Error> {
    let class = native.define_class("Image", ruby.class_object())?;
    class.define_singleton_method("get", function!(get, 1))?;
    class.define_singleton_method("list", function!(list, 0))?;
    class.define_singleton_method("inspect", function!(inspect, 1))?;
    class.define_singleton_method("remove", function!(remove, 2))?;
    class.define_singleton_method("prune", function!(prune, 0))?;
    Ok(())
}
